specs/Deployment/MSIXPackages.md

```diff
@@ -0,0 +1,189 @@
+This spec outlines the list of packages and their rules and roles.
+
+# MSIX Packages
+
+Windows App SDK is distributed as multiple MSIX packages:
+
+1. Framework
+2. Main
+3. Singleton
+4. Dynamic Dependency Lifetime Manager (DDLM)
+
+## Roles
+
+### Role - Framework
+
+The Framework package is the API delivery vehicle for Windows App SDK. The vast majority of code in
+the project is found here.
+
+Framework packages support concurrent behavior. A user may have multiple versions of a framework
+package in use across processes (process1 using 1.0.1.0, process2 using 1.0.2.0, ...). Updates can
+be registered for a user without disrupting running apps. Windows points running apps to updated
+Framework packages non-disruptively (e.g. when an app's not running), colaescing apps to newer
+versions over time (not necessarily instaneously). This provides for a very flexible servicing
+model.
+
+### Role - Main
+
+The Main package supplements the Framework package, for when functionality is needed but can't be
+delivered via the Framework package. Mechanisms include [Packaged COM](https://blogs.windows.com/windowsdeveloper/2017/04/13/com-server-ole-document-support-desktop-bridge/),
+[app services](https://docs.microsoft.com/windows/uwp/launch-resume/how-to-create-and-consume-an-app-service),
+[app extensions](https://docs.microsoft.com/uwp/api/Windows.ApplicationModel.AppExtensions.AppExtension),
+[background tasks](https://docs.microsoft.com/windows/uwp/launch-resume/create-and-register-a-background-task),
+[startup tasks](https://docs.microsoft.com/uwp/api/Windows.ApplicationModel.StartupTask) executing
+on user login, [ApplicationData](https://docs.microsoft.com/uwp/api/Windows.Storage.ApplicationData)
+for data storage scoped to Windows App SDK (not any particular app using Windows App SDK), and
+more.
+
+The primary scenarios include:
+
+* Brokered Access -- processes running in an [AppContainer](https://docs.microsoft.com/windows/win32/secauthz/appcontainer-isolation) requiring
+                     access to resources that cannot be access from within the AppContainer can use
+                     a 'helper process' running at MediumIL and doing work on the API's behalf.
+                     cannot do for itself (i.e. brokered access to a resource).
+* Long-Running Process (aka user service) -- functionality requiring a long-running process
+                                             (especially if its lifetime isn't directly coupled to
+                                             an app's lifetime).
+
+Main packages don't support side-by-side behavior for a user. A user can only have 1 (or 0) Main
+packages in a package family registered at a time. Thus if a user has v1.0.0.0 registered when
+1.0.0.1 is registered it's an *update*, ending with the user having v1.0.0.1 registered instead of
+v1.0.0.0. This poses servicing challenges, as updating the Main package requires any running
```

Comment:
What happens when an app built with 1..0.0.1 is trying to be installed, and the current Main package is 1.0.0.0, and there's currently an app open that uses Main package? What kind of error will the user get when they try to install? This seems slightly worrisome as we get more and more apps? How can developers "only rely on Main package when necessary", will it somehow be detected that an app isn't actively calling APIs that use Main even though the app itself requires Main?

I'd love to understand the user impact a bit more :) Great details already though!

Reply:
> What happens when an app built with 1..0.0.1 is trying to be installed, and the current Main package is 1.0.0.0, and there's currently an app open that uses Main package?

Very much depends on how features are designed and implemented, how they interact with Main package content, how termination and (re)start of those processes is handled and how all that interacts with applications (if at all).

The installer (currently) calls `PackageManager.AddPackageAsync(...,DeploymentOptions::None)`. Thus running Main processes won't be interrupted and the Main package won't be updated. For those who skip the installer and directly install the MSIX packages via `PackageManager` it depends if they do likewise or specify a 'force' (e.g. `DeploymentOptions::ForceTargetApplicationShutdown`). In the latter case it's no different than if you update Calculator or other packages while their app(s) are running - if they're implemented to provide Suspend/Resume semantics it looks to a user like the app disappears then reappears. YMMV of course.

>This seems slightly worrisome as we get more and more apps?

It's a pickle no matter what we do. If no Force option is specified then running apps are uninterrupted but a main package won't be updated. OTOH if a Force option's specified it may affect running applications. In the end this isn't all that different from installing various applications, some will interrupt functionality, require me to quit apps to unblock an uninstall, require me to reboot, etc. Servicing is a hard problem space to begin with, doing so generically for all possible software even harder, and doing it seamlessly with no impact is the holy grail of servicing. Balancing those factors along (without servicing Windows downlevel) makes for an imperfect equation.

This is one of the reasons we strongly encourage Windows App SDK developers to put content into the Framework package as it has no such servicing headaches.

>How can developers "only rely on Main package when necessary"

That's guidance to Windows App SDK developers (users of Windows App SDK won't care, to them it's just how the sausage is works).

Some features can accomplish their ends with different designs; even if harder / higher cost to implement a feature it's often the better solution in the end. We've seen some cases it's simply not practical (or even possible) to avoid; in such necessary cases, it's available.